rpc/kate-rpc/src/lib.rs

```diff
@@ -7,39 +7,41 @@ use codec::{Compact, Decode, Encode, Error as DecodeError, Input};
 use da_primitives::{
 	asdr::{AppExtrinsic, GetAppId},
 	traits::ExtendedHeader,
+	DataProof,
 };
-use frame_system::limits::BlockLength;
+use frame_support::traits::ExtrinsicCall;
+use frame_system::{limits::BlockLength, submitted_data};
 use jsonrpc_core::{Error as RpcError, Result};
 use jsonrpc_derive::rpc;
-use kate::{BlockDimensions, BlsScalar, PublicParameters};
+use kate::{com::Cell, BlockDimensions, BlsScalar, PublicParameters};
 use kate_rpc_runtime_api::KateParamsGetter;
 use lru::LruCache;
-use rs_merkle::{algorithms::Sha256, Hasher, MerkleTree};
 use sc_client_api::{BlockBackend, StorageProvider};
 use sp_api::ProvideRuntimeApi;
 use sp_blockchain::HeaderBackend;
-use sp_core::H256;
-use sp_rpc::number::NumberOrHex;
 use sp_runtime::{
 	generic::BlockId,
-	traits::{Block as BlockT, Header, NumberFor},
+	traits::{Block as BlockT, Extrinsic, Header},
 	AccountId32, MultiAddress, MultiSignature,
 };
 
+pub type HashOf<Block> = <Block as BlockT>::Hash;
+pub type CallOf<Block> = <<Block as BlockT>::Extrinsic as Extrinsic>::Call;
+
 #[rpc]
-pub trait KateApi {
+pub trait KateApi<Block, SDFilter>
+where
+	Block: BlockT,
+	SDFilter: submitted_data::Filter<CallOf<Block>>,
+{
 	#[rpc(name = "kate_queryProof")]
-	fn query_proof(
-		&self,
-		block_number: NumberOrHex,
-		cells: Vec<kate::com::Cell>,
-	) -> Result<Vec<u8>>;
+	fn query_proof(&self, cells: Vec<Cell>, at: Option<HashOf<Block>>) -> Result<Vec<u8>>;
 
 	#[rpc(name = "kate_blockLength")]
-	fn query_block_length(&self) -> Result<BlockLength>;
+	fn query_block_length(&self, at: Option<HashOf<Block>>) -> Result<BlockLength>;
 
 	#[rpc(name = "kate_queryDataProof")]
-	fn query_data_proof(&self, block_number: NumberOrHex, index: u32) -> Result<Vec<H256>>;
+	fn query_data_proof(&self, data_index: u32, at: Option<HashOf<Block>>) -> Result<DataProof>;
```

Comment:
does this mean the API now accepts both block number or hash instead of block number earlier?

Reply:
No, it only accepts the `hash`. It is a kind of de-facto standard in RPC calls on substrate.